### PR TITLE
LmP SBOMs

### PR DESCRIPTION
**Pull request by doanac**
These changes result in OE and Container builds both producing SBOMs in a consistent location so that we can start to build tooling to share them with users.

### CODE REVIEW

**Comment by doanac** [comment]
@mike-sul @ricardosalveti - this is still a little bit "work in progress". I'd like to work on the ota-lite api to make sure this approach works okay, but I don't foresee much, if anything, changing with this. So - now would be a good time to say this is a good/bad approach

**Review by ricardosalveti — APPROVED** [review]
LGTM

**Comment by ricardosalveti** [comment]
Having under sbom sounds good and it is also a fixed location.

File: apps/build.sh [hunk 1]
```diff
@@ -223,6 +223,11 @@ for x in $IMAGES ; do
 		echo "osftoken not provided, skipping publishing step"
 	fi
 
+	status "Doing a syft SBOM scan"
+	sbom_dst=/archive/sboms/${ct_base}/${ARCH}.sdpx.json
+	mkdir -p $(dirname $sbom_dst)
+	syft ${ct_base}:$TAG-$ARCH -o spdx-json > $sbom_dst
```
Comment (mike-sul) [inline apps/build.sh:229]: How long does it take to scan an image? If it takes some time I wonder if it's possible to output some progress.

File: apps/compose_apps.py [hunk 2]
```diff
@@ -37,8 +37,9 @@ def __init__(self, name, app_dir, image_downloader_cls=DockerDownloader):
 
             self._image_downloader_cls = image_downloader_cls
 
-            args = [self.DockerComposeTool, 'compose', '-f', self.ComposeFile, 'config']
-            cmd_exe(*args, cwd=self.dir)
+            if self.DockerComposeTool:
```
Comment (mike-sul) [inline apps/compose_apps.py:40]: Is it temporal change or is it done for some purpose?

Comment (mike-sul) [reply apps/compose_apps.py:40]: I see now why you did it.

File: apps/build.sh (hunk 1 above)
Comment (doanac) [reply apps/build.sh:229]: it sends progress to stderr

File: apps/generate_non_factory_sboms.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    # We don't need to validate the compose content here - just need to
+    # iterate service images.
+    ComposeApps.App.DockerComposeTool = None
+
+    apps = ComposeApps(args.apps_root)
+    analyzed = {}
+    for app in apps:
+        for img in app.images():
+            img = expandvars(img)
```
Comment (mike-sul) [inline apps/generate_non_factory_sboms.py:24]: Maybe it makes sense to move it to `ComposeApps.Apps.images`.

File: apps/generate_non_factory_sboms.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+from helpers import status
+
+
+def main(args):
+    status("Loading docker-compose files to find non-factory containers")
+    factory_prefix = f"hub.foundries.io/{args.factory}/"
+
+    # We don't need to validate the compose content here - just need to
+    # iterate service images.
+    ComposeApps.App.DockerComposeTool = None
```
Comment (mike-sul) [inline apps/generate_non_factory_sboms.py:18]: Is validation causes any issues or there was another reason why you turned it off?

Comment (doanac) [reply apps/generate_non_factory_sboms.py:18]: it gets noisy when you see the CI output. it starts dumping the compose content to stdout and doesn't make much sense why

**Comment by mike-sul** [comment]
I wonder whether it makes sense to do scanning for both hub.f.io and 3rd party hosted images within the same script. It looks like implementation will be encapsulated in one spot in this case.

File: apps/generate_non_factory_sboms.py (hunk 4 above)
Comment (mike-sul) [reply apps/generate_non_factory_sboms.py:18]: Probably we need to add `--quite` flag` to the `config` cmd.

**Comment by doanac** [comment]
> I wonder whether it makes sense to do scanning for both hub.f.io and 3rd party hosted images within the same script. It looks like implementation will be encapsulated in one spot in this case.

I thought about that and actually implemented that. However, it's inefficient. The publish script has to download all the per-arch container images that we'd built earlier.

File: apps/generate_non_factory_sboms.py (hunk 3 above)
Comment (doanac) [reply apps/generate_non_factory_sboms.py]: https://github.com/foundriesio/ci-scripts/pull/208/commits/892c7a21f8571d4d1b523dda05f7d88502787468

File: apps/generate_non_factory_sboms.py (hunk 4 above)
Comment (doanac) [reply apps/generate_non_factory_sboms.py]: https://github.com/foundriesio/ci-scripts/pull/208/commits/19c68c605664807fc8dca2905a2752d8eb8d3657

**Comment by doanac** [comment]
@mike-sul - good thing we are re-testing this. the move to `docker compose` has broken my change. I need to pull golang compose into https://github.com/foundriesio/docker-dind

File: apps/compose_apps.py [hunk 5]
```diff
@@ -38,17 +41,23 @@ def __init__(self, name, app_dir, image_downloader_cls=DockerDownloader):
             self._image_downloader_cls = image_downloader_cls
 
             args = [self.DockerComposeTool, 'compose', '-f', self.ComposeFile, 'config']
-            cmd_exe(*args, cwd=self.dir)
+            if quiet:
```
Comment (mike-sul) [inline apps/compose_apps.py:44]: I think that's fine. I just thought about another option

```
if quite:
  args.append('--quite')

```

Comment (mike-sul) [reply apps/compose_apps.py:44]: > The publish script has to download all the per-arch container images

I thought about it too. It's just not necessary, just thought, we know a tag of images we build, i.e. < commit short hash>.

**Comment by mike-sul** [comment]
> I need to pull golang compose into https://github.com/foundriesio/docker-dind

**Comment by mike-sul** [comment]
@doanac I just closed this PR by mistake, I reopened it.

File: apps/generate_non_factory_sboms.py (hunk 4 above)
Comment (doanac) [reply apps/generate_non_factory_sboms.py]: i've done another force push to use your suggestion and have a new ci container. everything should be in place now

**Review by mike-sul — APPROVED** [review]

**Pull request merged**